### ISSUE TITLE
Causes Issues With Facebook Module

### ISSUE THREAD

**Issue by tennist** [signalpoint/legal#1]
I have this module and the facebook module installed in my DrupalGap app.  When I try to create an account using facebook everything works fine server side and the account is created , but nothing happens on the app side and the login page is reloaded.  I believe this has to due with the fact that the user is not given the option to accept the terms of service.  Any ideas on how to bring up the terms of service after the user creates an account with facebook?

**Comment by signalpoint** [comment]
@tennist Which terms of service? Perhaps the Legal module in Drupal? From recollection, there were no T&C's on the Facebook side (aside from what folks agree to when first creating their Facebook account on FB).

**Comment by signalpoint** [comment]
@tennist Whoops, I see this is in the `legal` module queue... so yes, you are indeed referring to the Legal module. I haven't been able to expand this far enough yet to actually support the acceptance of the T&C in DrupalGap. Instead, I've been bypassing it in the app by implementing `hook_form_alter()` in *Drupal*, and removing the `legal` element using an `#after_build` from the form, e.g.:

```
/**
 * Implements hook_form_alter().
 */
function examp[le_form_alter(&$form, &$form_state, $form_id) {
  if ($form_id == 'user_register_form' && arg(0) == 'drupalgap') {
    $form['#after_build'][] = 'example_form_after_build';
  }
}

function example_form_after_build($form, $form_state) {
  // Skip legal requirements on user registration form.
  if (isset($form['legal'])) { unset($form['legal']); }
  return $form;
}
```

As a reminder, since the Sevices module utilizes the Drupal Forms API for many of the core service calls, we can use form alterations on the service calls.

You might be able to do it directly in `hook_form_alter()`, instead of having to use the `#after_build`.

**Comment by tennist** [comment]
In setting up the legal module for drupalgap, there is a patch required for the legal module on the sever side.  This patch creates a setting for the legal module to ignore the T&C acceptance for a list of login paths The login path drupalgap/user/login.json is added automatically as part of the patch.  I was able to add drupalgap/fboauth/connect.json to this list to get my app to function with facebook login.  One big issue I see here is that my users will never be supplied with the terms & conditions if they use the app to create an account using facebook.  Everytime they login the check on the T&C flag is bypassed.  

The same sort of issue arises if they login to the app normally (username and password).  In this route they must accept the original T&C to create an account, however, the user will never be presented with any updates that I make to the T&C since the check for acceptance of the latest T&C is bypassed due to drupalgap/user/login.json being in the list.  Basically the legal module for drupal gap only offers the ability to accept T&C on original account creation by standard registration.  

I think what is being done through the legal module on the server side can be implemented on the app side  The two bypasses would need to remain in place on the server to allow the user info to be passed to the app, but then the app needs to make a call to the server to check if the latest T&C has been accepted.  On the server side, the legal module uses hook_user_login() to check that the user has accepted the latest T&C upon login.  If not, it logs the user out ,stores their login info in the background, and presents them with the T&C form.  Once they accept them and click confirm, it uses the stored login info to log them in.

Is there a similar function to the Drupal hook_user_login() in Drupalgap?  I am not finding one, but perhaps you know of a way to mimic it.  If so I will try to make a patch for the legal module in Drupalgap that does the same checks.

**Comment by signalpoint** [comment]
@tennist Agreed. I think this would be best accomplished by using `hook_services_request_postprocess_alter()` in Drupal (the DrupalGap Legal module up in Drupal): http://cgit.drupalcode.org/services/tree/docs/services.alter.api.php?h=7.x-3.x

With that, we'd

1. make calls to the Legal API/functions to determine if the current logged in user has accepted the terms and conditions (as well as the latest terms and conditions)
2. append that data in a way that DrupalGap's Legal module (down in the SDK) would use hook services post process on the System Connect call
3. if it noticed that they hadn't accepted the terms and conditions, it'd redirect them to a page in the app (again provided by DrupalGap Legal module in the SDK) where it'd display the T&C and let them accept it (a custom form)
4. upon submission would need to make a custom Service Call back up to Drupal (the Resource provided up in Drupal, and the Service call down in the SDK) to mark their account as current.

Easier said than done, but it should be do-able. Thoughts?

**Comment by tennist** [comment]
@signalpoint  Thanks for the followup on this.  I had already started down the road you mentioned using the hook_service_request_postprocess_alter() over the weekend.  I was able to accomplish what I wanted on the drupalgap side using using the hook_post_process_page().  I just have a bit of code cleanup to perform and will submit the revised module on both sides.

**Comment by tennist** [comment]
@signalpoint  Attached is the updated legal module for drupalgap and the updated drupalgap_legal module for drupal.  The module now checks that a user has accepted the latest terms and conditions once they log in and try to navigate to any page.  If they have not, it redirects them to a form to accept the latest terms and conditions.
  
[legal.zip](https://github.com/signalpoint/legal/files/739782/legal.zip)
[drupalgap_legal.zip](https://github.com/signalpoint/legal/files/739790/drupalgap_legal.zip)

**Comment by luxio** [comment]
@tennist Thank you for sharing your code.

Does your DG module sends the information, that the user has accepted the latest term and conditions back to Drupal?

**Comment by signalpoint** [comment]
This may be helpful too: https://www.drupal.org/project/services_legal

**Comment by luxio** [comment]
@signalpoint Thank you for sharing this. Is there a DG Module wich integrates the services_legal module?

**Comment by signalpoint** [comment]
@luxio No, there is no corresponding DG module. The services_legal module is for those applications implementing their own System Connect and User Authentication procedures, and it allows them to easily see the user's legal status.

**Comment by luxio** [comment]
@signalpoint Thanks.
